File: gitstatus.py
#!/usr/bin/env python
from __future__ import print_function
from subprocess import Popen, PIPE

# change this symbol to whatever you prefer
prehash = ':'

# checking if i am inside git work tree.
# git rev-parse --is-inside-work-tree
gitsym = Popen(['git', 'rev-parse', '--is-inside-work-tree'], stdout=PIPE, stderr=PIPE)
error = gitsym.communicate()[1]

# code 128: git did not exit cleanly (exit code 128).
# in this case, it means that there is no ".git" directory.
if (gitsym.returncode == 128):
	gitsym.kill()
	exit(1)

error_string = error.decode('utf-8')
if 'fatal: Not a git repository' in error_string: 
    gitsym.kill()
    exit(1)

res, err = Popen(['git','diff','--name-status'], stdout=PIPE, stderr=PIPE).communicate()
err_string = err.decode('utf-8')
if 'fatal' in err_string: 
        gitsym.kill()   
        exit(1)

# get name of current branch
# git rev-parse --abbrev-ref HEAD
branch = Popen(['git', 'rev-parse', '--abbrev-ref', 'HEAD'], stdout=PIPE).communicate()[0].decode().strip()

# gets number of staged files
# git diff --cached --numstat | wc -l
staged = len(Popen(['git', 'diff', '--cached', '--numstat'], stdout=PIPE).communicate()[0].decode().strip().splitlines())

# gets number of conflicts
# git --no-pager diff --name-only --diff-filter=U | wc -l 
conflicts = len(Popen(['git', '--no-pager', 'diff', '--name-only', '--diff-filter=U'], stdout=PIPE).communicate()[0].decode().strip().splitlines())

# gets number of modified files
# git --no-pager diff --name-only --diff-filter=M | wc -l
modified = len(Popen(['git', '--no-pager', 'diff', '--name-only', '--diff-filter=M'], stdout=PIPE).communicate()[0].decode().strip().splitlines())

# gets number of untracked files
# git ls-files --others --exclude-standard | wc -l
untracked = len(Popen(['git', 'ls-files', '--others', '--exclude-standard'], stdout=PIPE).communicate()[0].decode().strip().splitlines())

# gets number of deleted files
# git --no-pager diff --name-only --diff-filter=D | wc -l
deleted = len(Popen(['git', '--no-pager', 'diff', '--name-only', '--diff-filter=D'], stdout=PIPE).communicate()[0].decode().strip().splitlines())

ahead, behind = 0,0

if not branch: # not on any branch
	branch = prehash + Popen(['git','rev-parse','--short','HEAD'], stdout=PIPE).communicate()[0].decode("utf-8")[:-1]
else:
	remote_name = Popen(['git','config','branch.%s.remote' % branch], stdout=PIPE).communicate()[0].decode("utf-8").strip()
	if remote_name:
		merge_name = Popen(['git','config','branch.%s.merge' % branch], stdout=PIPE).communicate()[0].decode("utf-8").strip()
		if remote_name == '.': # local
			remote_ref = merge_name
		else:
			remote_ref = 'refs/remotes/%s/%s' % (remote_name, merge_name[11:])
		revgit = Popen(['git', 'rev-list', '--left-right', '%s...HEAD' % remote_ref],stdout=PIPE, stderr=PIPE)
		revlist = revgit.communicate()[0]
		if revgit.poll(): # fallback to local
			revlist = Popen(['git', 'rev-list', '--left-right', '%s...HEAD' % merge_name],stdout=PIPE, stderr=PIPE).communicate()[0]
		behead = revlist.decode("utf-8").splitlines()
		ahead = len([x for x in behead if x[0]=='>'])
		behind = len(behead) - ahead

out = ' '.join([
	branch,
	str(ahead),
	str(behind),
	str(staged),
	str(conflicts),
	str(modified),
	str(untracked),
	str(deleted)
	])
print(out, end='')

File: gitstatus.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int run_command(const char *cmd, char *result, int max_len) {
  FILE *fp;
  fp = popen(cmd, "r");
  if (fp == NULL) {
    return -1;
  }
  if (fgets(result, max_len, fp) == NULL) {
    pclose(fp);
    return -1;
  }
  pclose(fp);
  // Remove trailing newline
  result[strcspn(result, "\n")] = 0;
  return 0;
}

int main() {
  char result[1024];

  // Check if inside a git repository
  if (run_command("git rev-parse --is-inside-work-tree 2>/dev/null", result,
                  sizeof(result)) != 0 ||
      strcmp(result, "true") != 0) {
    return 1;
  }

  // Get current branch name
  if (run_command("git rev-parse --abbrev-ref HEAD", result, sizeof(result)) !=
      0) {
    return 1;
  }
  char branch[256];
  strcpy(branch, result);

  // Get number of staged files
  run_command("git diff --cached --numstat | wc -l", result, sizeof(result));
  int staged = atoi(result);

  // Get number of conflicts
  run_command("git --no-pager diff --name-only --diff-filter=U | wc -l", result,
              sizeof(result));
  int conflicts = atoi(result);

  // Get number of modified files
  run_command("git --no-pager diff --name-only --diff-filter=M | wc -l", result,
              sizeof(result));
  int modified = atoi(result);

  // Get number of untracked files
  run_command("git ls-files --others --exclude-standard | wc -l", result,
              sizeof(result));
  int untracked = atoi(result);

  // Get number of deleted files
  run_command("git --no-pager diff --name-only --diff-filter=D | wc -l", result,
              sizeof(result));
  int deleted = atoi(result);

  // Get ahead and behind counts
  int ahead = 0, behind = 0;
  char remote_name[256], merge_name[256], remote_ref[256];

  // Check if on a specific branch or detached HEAD
  if (strcmp(branch, "HEAD") == 0) {
    run_command("git rev-parse --short HEAD", result, sizeof(result));
    snprintf(branch, sizeof(branch), ":%s", result);
  } else {
    run_command("git config branch.%s.remote", remote_name,
                sizeof(remote_name));
    run_command("git config branch.%s.merge", merge_name, sizeof(merge_name));

    if (strcmp(remote_name, ".") == 0) {
      snprintf(remote_ref, sizeof(remote_ref), "%s", merge_name);
    } else {
      snprintf(remote_ref, sizeof(remote_ref), "refs/remotes/%s/%s",
               remote_name, merge_name + 11);
    }

    snprintf(result, sizeof(result), "git rev-list --left-right %s...HEAD",
             remote_ref);
    FILE *revlist_fp = popen(result, "r");
    while (fgets(result, sizeof(result), revlist_fp) != NULL) {
      if (result[0] == '>') {
        ahead++;
      } else {
        behind++;
      }
    }
    pclose(revlist_fp);
  }

  printf("%s %d %d %d %d %d %d %d\n", branch, ahead, behind, staged, conflicts,
         modified, untracked, deleted);
  return 0;
}
